upgrade_all_content.php

```diff
@@ -0,0 +1,134 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Upgrade all content to latest libraries script.
+ *
+ * @package   mod_hvp
+ * @author    Rossco Hellmans <rosscohellmans@catalyst-au.net>
+ * @copyright Catalyst IT, 2021
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once("../../config.php");
+require_once($CFG->libdir.'/adminlib.php');
+require_once("locallib.php");
+
+// No guest autologin.
+require_login(0, false);
+
+$confirm = optional_param('confirm', null, PARAM_INT);
+
+$returnurl = new moodle_url('/mod/hvp/library_list.php');
+$pageurl = new moodle_url('/mod/hvp/update_libraries.php');
+$PAGE->set_url($pageurl);
+
+admin_externalpage_setup('h5plibraries');
+
+$PAGE->set_title("{$SITE->shortname}: " . get_string('libraries', 'hvp'));
+
+if ($confirm && confirm_sesskey()) {
+
+    // We may need extra execution time and memory.
+    core_php_time_limit::raise(HOURSECS);
+    raise_memory_limit(MEMORY_EXTRA);
+
+    $editor = mod_hvp\framework::instance('editor');
+    $ajax = $editor->ajax;
+    $core = $ajax->core;
+
+    // Grab all the libraries and grab any that can be upgraded.
+    $settings = ['libraryInfo' => []];
+    $libraries = $ajax->core->h5pF->loadLibraries();
+    foreach ($libraries as $versions) {
+        foreach ($versions as $library) {
+            $restricted = isset($library->restricted) && $library->restricted == 1;
+            if (!$restricted && $library->runnable) {
+                $upgrades = $core->getUpgrades($library, $versions);
+                $numcontents = $core->h5pF->getNumContent($library->id);
+                if (!empty($upgrades) && $numcontents > 0) {
+                    $fromver = $library->major_version . '.' . $library->minor_version . '.' . $library->patch_version;
+                    $tover = end($upgrades);
+                    $a = (object)[
+                        'from' => $library->title . ' (' . $fromver . ')',
+                        'to' => $library->title . ' (' . $tover . ')',
+                        'count' => $numcontents,
+                    ];
+                    $settings['libraryInfo'][] = [
```

Comment:
Can we move out the non-dynamic stuff then just copy and modify the non static fields in the loop? Saves a ton of string loads, and if there are a lot of libs, that might overrun the string cache, it only has room for 2000 strings in APC by default

Reply:
Good idea, I've moved all of the non dynamic stuff into it's own separate array and adjusted the js to use that